### PR TITLE
Add date_add function.

### PR DESCRIPTION
**Pull request by qiuyesuifeng**
1 move extractTime to time.go and rename to extractTimeNum.
2 add date_add function.
3 to pass mybatis test case.

### CODE REVIEW

File: expression/date_add_test.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	. "github.com/pingcap/check"
+	mysql "github.com/pingcap/tidb/mysqldef"
+)
+
+var _ = Suite(&testDateAddSuite{})
+
+type testDateAddSuite struct {
+}
+
+func (t *testDateAddSuite) TestExtract(c *C) {
```
Comment (shenli) [inline expression/date_add_test.go]: Why the function name is "TestExtract"?

Comment (qiuyesuifeng) [inline expression/date_add_test.go]: My fault, copied this test case and forgot to change function name.

File: expression/visitor.go [hunk 2]
```diff
@@ -478,3 +481,19 @@ func (bv *BaseVisitor) VisitFunctionTrim(ss *FunctionTrim) (Expression, error) {
...
+// VisitDateAdd implements Visitor interface.
+func (bv *BaseVisitor) VisitDateAdd(da *DateAdd) (Expression, error) {
+	var err error
+	da.Date, err = da.Date.Accept(bv.V)
+	if err != nil {
+		return da, errors.Trace(err)
+	}
+
+	da.Interval, err = da.Interval.Accept(bv.V)
+	if err != nil {
```
Comment (shenli) [inline expression/visitor.go]: return da, errors.Trace(err)

File: parser/parser.y [hunk 3]
```diff
@@ -2228,6 +2229,14 @@ FunctionCallNonKeyword:
 			return 1
 		}
 	}
+|	"DATE_ADD" '(' Expression ',' "INTERVAL" Expression TimeUnit ')'
```
Comment (shenli) [inline parser/parser.y]: date_add is not a keyword. Add it to NoKeywordToken list.

File: expression/visitor.go (hunk 2 above)
Comment (qiuyesuifeng) [inline expression/visitor.go]: i think 
```
if err != nil {
     return errors.Trace(err)
}

return nil
```
will be much clearer though have to add more code.

File: mysqldef/fsp.go [hunk 4]
```diff
@@ -80,3 +80,13 @@ func parseFrac(s string, fsp int) (int, error) {
 	//  0.0312 round 2 -> 3 -> 30000
 	return int(round * math.Pow10(MaxFsp-fsp)), nil
 }
+
+// alignFrac is used to generate alignment frac, like `100` -> `100000`
+func alignFrac(s string, fsp int) string {
```
Comment (siddontang) [inline mysqldef/fsp.go]: how about len(s) > fsp?

**Comment by siddontang** [comment]
You can use date expression instead, we will use for date sub or other later.

**Comment by qiuyesuifeng** [comment]
Aggreed, i will do it in next PR.

**Comment by siddontang** [comment]
Got it.
LGTM

File: mysqldef/fsp.go (hunk 4 above)
Comment (qiuyesuifeng) [inline mysqldef/fsp.go]: it will be treated as number, like `10000000`,  means 10000000 microseconds, but `100` means 100000 microseconds.

**Comment by shenli** [comment]
LGTM

**Pull request merged**